tortoise/util.go

```diff
@@ -115,16 +119,84 @@ func persistContextualValidity(logger log.Log,
 	blocks map[types.LayerID][]types.BlockID,
 	opinion votes,
 ) error {
-	for lid := from.Add(1); !lid.After(to); lid = lid.Add(1) {
+	var err error
+	iterateLayers(from.Add(1), to, func(lid types.LayerID) bool {
 		for _, bid := range blocks[lid] {
 			sign := opinion[bid]
 			if sign == abstain {
 				logger.With().Panic("bug: layer should not be verified if there is an undecided block", lid, bid)
 			}
-			if err := bdp.SaveContextualValidity(bid, lid, sign == support); err != nil {
-				return fmt.Errorf("saving validity for %s: %w", bid, err)
+			err = bdp.SaveContextualValidity(bid, lid, sign == support)
+			if err != nil {
+				err = fmt.Errorf("saving validity for %s: %w", bid, err)
+				return false
 			}
 		}
+		return true
+	})
+	return err
+}
+
+func iterateLayers(from, to types.LayerID, callback func(types.LayerID) bool) {
+	for lid := from; !lid.After(to); lid = lid.Add(1) {
+		if !callback(lid) {
+			return
+		}
 	}
-	return nil
+}
+
+type layerVerifier interface {
+	verify(log.Log, types.LayerID) bool
+}
+
+type voteReason string
+
+func (v voteReason) String() string {
+	return string(v)
+}
+
+const (
+	reasonHareOutput     voteReason = "hare"
+	reasonValiditiy      voteReason = "validity"
+	reasonLocalThreshold voteReason = "local_threshold"
+	reasonCoinflip       voteReason = "coinflip"
+)
+
+// lsb is a mode, second bit is rerun
+// examples (lsb is on the right, prefix with 6 bits is droppped)
+// 10 - rerun in verifying
+// 00 - live tortoise in verifying
+// 11 - rerun in full
+// 01 - live tortoise in full.
+type mode uint8
```

Comment:
is this performance critical that requires bit manipulation like this? what's wrong with the previous version with iota?

Reply:
the reason was that i can change rerun state without changing mode, and mode without changing rerun. using 1 and 2 bit was convenient, so i can toggle them in toggleRerun and toggleMode. i can use [2]bool instead if that's seems odd